sdk/servicebus/service-bus/src/core/linkEntity.ts

```diff
@@ -196,20 +205,30 @@ export abstract class LinkEntity<LinkT extends Receiver | AwaitableSender | Requ
     return result;
   }
 
-  /**
-   * Indicates that a link initialization is in process.
-   */
-  get isConnecting(): boolean {
-    return this._isConnecting;
-  }
-
   /**
    * Initializes this LinkEntity, setting this._link with the result of  `createRheaLink`, which
    * is implemented by child classes.
    *
    * @returns A Promise that resolves when the link has been properly initialized
    */
   async initLink(options: LinkOptionsT<LinkT>, abortSignal?: AbortSignalLike): Promise<void> {
+    // we'll check that the connection isn't in the process of recycling (and if so, wait for it to complete)
+    log.error(`${this._logPrefix} Checking if connection is ready`);
+    await this._context.readyToOpenLink();
```

Comment:
Any reason this is done before the openLock is acquired instead of after?

Reply:
There is a deadlock scenario because of how much work 'readyToOpenLink' can do.

Basically, you can end up in this scenario:

(this was the prior code)

```
lockEntity.initLink()
  <takes lock>
  lockEntity.negotiateClaim()
    context.readyToOpenLink()
      context.refreshConnection()
        lockEntity.closeLink()
          <tries to take lock but it's already held>
```

So I fixed this by moving the contentious part out of the lock and splitting what I was doing into two pieces:
1. The first piece (the readyForOpenLink) happens outside of the lock. This makes it safe for external callers (like ConnectionContext) to do link manipulation operations and also gives me some reassurance that the link is at least up.
2. In negotiateClaim rather than call readyForOpenLink I do the non-calling-me-reentrantly check with context.isConnectionClosing() to make sure that the link isn't in an indeterminate state. This should be safe for me to call anytime.